tsz-compress/src/v2/mod.rs

```diff
@@ -1,18 +1,1355 @@
+            }
+            // ZigZag decoding
+            let value = (value >> 1) ^ -(value & 1);
+
+            decoded_buffer[decoded_buffer_index] = value as i8;
+            decoded_buffer_index += 1;
+            idx += 9;
+        } else {
+            return Err("Invalid encoding for i8");
+        }
+    }
+    if idx == bits.len() {
+        Ok(decoded_buffer)
+    } else {
+        return Err("Some value remains to be decoded.");
+    }
+}
+
+pub fn decode_delta_delta_i16(bits: &'_ BitBufferSlice) -> Result<[i16; 10], &'static str> {
+    if bits.is_empty() {
+        return Err("Not enough bits to decode");
+    }
+    let mut decoded_buffer: [i16; 10] = [0; 10];
+    let mut decoded_buffer_index = 0;
+    let mut idx = 0;
+    while idx < bits.len() {
+        if bits[idx] {
+            return Err("Not in delta-delta compressed. Decode with delta decoding.");
+        }
+        let mut value = 0;
+        if !bits[idx + 1] && !bits[idx + 2] {
+            idx += 3;
+            value = bits[idx] as i16;
+            value = (value >> 1) ^ -(value & 1);
+            decoded_buffer[decoded_buffer_index] = value;
+            decoded_buffer_index += 1;
+            idx += 1;
+        } else if !bits[idx + 1] && bits[idx + 2] {
+            idx += 3;
+            for i in 0..5 {
+                value |= (bits[idx + i] as i16) << i;
+            }
+            value = (value >> 1) ^ -(value & 1);
+            decoded_buffer[decoded_buffer_index] = value;
+            decoded_buffer_index += 1;
+            idx += 5;
+        } else if bits[idx + 1] && !bits[idx + 2] {
+            idx += 3;
+            for i in 0..9 {
+                value |= (bits[idx + i] as i16) << i;
+            }
+            value = (value >> 1) ^ -(value & 1);
+            decoded_buffer[decoded_buffer_index] = value;
+            decoded_buffer_index += 1;
+            idx += 9;
+        } else if bits[idx + 1] && bits[idx + 2] && !bits[idx + 3] {
+            idx += 4;
+            for i in 0..16 {
+                value |= (bits[idx + i] as i16) << i;
+            }
+            value = (value >> 1) ^ -(value & 1);
+            decoded_buffer[decoded_buffer_index] = value;
+            decoded_buffer_index += 1;
+            idx += 16;
+        } else if bits[idx + 1] && bits[idx + 2] && bits[idx + 3] {
+            idx += 4;
+            let mut value: i64 = 0;
+            for i in 0..64 {
+                value |= (bits[idx + i] as i64) << i;
+            }
+            value = (value >> 1) ^ -(value & 1);
+            decoded_buffer[decoded_buffer_index] = value as i16;
+            decoded_buffer_index += 1;
+            idx += 64;
+        } else {
+            return Err("Invalid encoding for i16");
+        }
+    }
+    if idx == bits.len() {
+        Ok(decoded_buffer)
+    } else {
+        return Err("Some value remains to be decoded.");
+    }
+}
+
+pub fn decode_delta_delta_i32(bits: &'_ BitBufferSlice) -> Result<[i32; 10], &'static str> {
```

Comment:
If decoding delta_delta bits from a stream that is mostly delta bits, then the decode delta and delta-delta need to share the same decoding scheme. That means we only need one decode_iN for i8, i16, i32, i64.

It makes sense to me for such a method to iterate over the whole column and directly extend into the resultant Vec. As we decompress, we will have to track the previous value and delta in order to reconstruct from the delta or delta-delta bits.

If we decode delta bits, then we would recover up to 10 elements from deltas per header. If we decode delta-delta bits, then we would recover one element per header.